### ISSUE TITLE
tlog server : proper handle of failed flush

### ISSUE THREAD

**Issue by iwanbk** [g8os/blockstor#131]
moved from https://github.com/g8os/tlog/issues/66.

put back tlogs to memory and retry it?

- [x] retry 3 times, if failed to store to redis
- [x] last_hash of a vdisk must only be updated if all pieces successfully stored (or use another data structure, see : https://github.com/g8os/blockstor/issues/131#issuecomment-301045314) (#154)

**Comment by GlenDC** [comment]
The question is of course how many retries do we want to allow? And are all failures equal, as in, should all types of failure trigger a retry, or are their fatal failures in which cases no retry would ever help.

Also, even though we keep them in memory, what if the service stops before those blocks in memory, were able to be sent to the tlogserver. Knowing this, should we also persist them on the client side (nbdserver) in some temporary cache file (or perhaps even better an ardb), or is this too much?

**Comment by iwanbk** [comment]
```
Knowing this, should we also persist them on the client side (nbdserver) in some temporary cache file (or perhaps even better an ardb
```
We at least need to keep the block in nbdserver memory until that block received by the tlog server.

**Comment by iwanbk** [comment]
updated first comment, write down the tasks list of this issue

**Comment by iwanbk** [comment]
```
 last_hash of a vdisk must only be updated if all pieces successfully stored
```
In #144, flusher will rollback the last_hash of vdisk if pieces failed to be stored.
But if still has issue if the rollback itself failed.

It is a disaster because without knowing correct last hash, we couldn't restore the entire block chain.

As an alternative solution, instead of only store the last hash, we can store last X of last hash.
So, if the last one failed, we can use last - 1.

I think storing few last hash will be enough.

**Comment by GlenDC** [comment]
> retry 3 times

This value looks pretty random and based on nothing. Also I stated before, I don't think it makes sense for all types of error to retry, as for some kind of failures you could retry all you want and it wouldn't make a difference.

TLDR: I think there should be a little more intelligence behind this, than simply retrying 3 times.

**Comment by GlenDC** [comment]
> We at least need to keep the block in nbdserver memory until that block received by the tlog server.

Sure, but what if the nbdserver restarts/stops for whatever reason, while there are still some blocks in memory. Is it ok for these to be lost? Or do we want to store them in persistent memory or another type of storage?

**Comment by iwanbk** [comment]
> This value looks pretty random and based on nothing. Also I stated before, I don't think it makes sense 
> for all types of error to retry, as for some kind of failures you could retry all you want and it wouldn't make > a difference.

updated to
```
retry 3 times, if failed to store to redis
```

It makes sense to retry if we failed to store to redis/ardb
- redis ardb crashed, we need to wait for it to up again
- for some reason, network connection closed, we need to retry 
- etc..

and yes, 3 times is a bit random, let me know if you have better idea....we can always improve it :)

**Comment by iwanbk** [comment]
```
Sure, but what if the nbdserver restarts/stops for whatever reason, while there are still some blocks in memory. Is it ok for these to be lost? Or do we want to store them in persistent memory or another type of storage?
```
What do you think the best for nbdserver?

**Comment by GlenDC** [comment]
At the moment I'm not sure what is best. All I know for now is that the nbdserver can stop before all tlog aggregations are stored. Not sure if that's fine or not.

**Comment by iwanbk** [comment]
About 

> last_hash of a vdisk must only be updated if all pieces successfully stored (or use another data structure, see

Instead of storing only one last hash, we can improve it by storing X last hash. X=2 should be enough, but make it bigger doesn't give any harm either.

Only storing one last hash will give us a problem if the last aggregation is failed to be stored to ardb but we already store the new last hash to ardb. We can of course try to roll it back, but the rollback itself might also failed.

To avoid the need to rollback, we can also only store the new last hash after all pieces successfully stored. But we are going to get performance hit.

So, storing X last hash is better choice here:
- in normal condition, what we really need is the very last hash.
- if the very last value is not valid, we can use last -1
- if last -1 still not valid, we can use last-2...and so on.

we need to keep X value to be small to avoid performance issue.

**Comment by iwanbk** [comment]
> At the moment I'm not sure what is best. All I know for now is that the nbdserver can stop before all tlog aggregations are stored. Not sure if that's fine or not.

I think the operation can only be flushed by nbdserver after it successfully flushed by tlog.
